vmm: live migration asynchronization (dedicated thread)

phip1611 opened the pull request:
The PR prepares the live-migration statistics. For that, we need to ensure that `SendMigration` won't block the whole VMM thread's event loop all the time ("asynchronization").

Further, this PR lays the groundwork to prevent VM change events (resize, hotplug) when a migration is ongoing.

## Changes
- EventFD logic to sync VMM from the VM Migration Thread
- Put migration into dedicated thread that takes ownership of the VM from the VMM
- Let all endpoints know when a migration is ongoing (the VMM knows the VMs ownership

**Please review this commit-by-commit.**


Closes https://github.com/cobaltcore-dev/cobaltcore/issues/293 and https://github.com/cobaltcore-dev/cobaltcore/issues/230


## Steps Before Merge
- [x] passes libvirt-tests
- [x] Team aggrees design

phip1611 committed:
vmm: better VM ownership handling in context of live migration

The commit prepares to properly handle API events during ongoing
live-migrations. The VmInfo call is currently not working when a VM is
migrating. This will be addressed in a follow-up as part of statistics
migration statistics about ongoing live-migrations.

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
vmm: api: temporarily make VmSendMigration call blocking again

Once we have a mechanism to query the progress of an ongoing
live-migration, we can remove this workaround.

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

amphi commented on vmm/src/lib.rs:
```diff
@@ -2529,11 +2770,35 @@ impl RequestHandler for Vmm {
+            return Err(MigratableError::MigrateSend(anyhow!(
+                "Another migration is already in progress"
+            )));
+        }
```
Is this still necessary with the `MaybeVmOwnership` enum?

phip1611 replied:
good catch

amphi commented on vmm/src/lib.rs:
```diff
@@ -2548,42 +2813,37 @@ impl RequestHandler for Vmm {
-                }
+        // Take VM ownership. This also means that API events can no longer
+        // change the VM (e.g. net device hotplug).
+        let vm = self.vm.take_vm();
```
When exactly do you put `MaybeVmOwnership::Migration` into `Vmm::vm`?

phip1611 replied:
good catch, I missed that 😁  we need a test soon that properly test for a proper response if a migration is ongoing

phip1611 committed:
misc: fix typo and wording

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

amphi commented on vmm/src/lib.rs:
```diff
@@ -772,6 +772,40 @@ pub struct VmmThreadHandle {
+pub enum MaybeVmOwnership {
+    /// The VMM holds the ownership of the VM.
+    Vmm(Vm),
+    /// The VM is temporarily by the current ongoing migration.
```
I somehow cannot suggest a change, but I guess you meant something like

The VM is temporarily **blocked** by the current ongoing migration.

phip1611 replied:
yes, thanks

> I somehow cannot suggest a change, but I guess you meant something like

This only works in the "total diff view", not in the "per commit diff view" :/

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -712,6 +715,70 @@ impl MigrationState {
+        self.sender.send(payload).unwrap();
+
+        // Trigger the event of checking migration results.
+        // VMM will take ownership of VM back.
```
nit:
```suggestion
// Notify the listener(s) to check migration results.
// The VMM will reclaim ownership of the VM.
```

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -739,6 +811,14 @@ pub struct Vmm {
+    ///
+    /// This enables to
+    /// - give back control of the VM to the VMM
+    /// - verify if the migration has been successful
```
nit:
```suggestion
    /// Receiving side of a channel to the migration thread.
    ///
    /// This enables
    /// - giving back control of the VM to the VMM
    /// - verifying whether the migration succeeded
```

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -712,6 +715,70 @@ impl MigrationState {
+    /// Sender that transfers ownership back to the VMM, once the migration is
+    /// finished.
+    sender: Sender<(Vm, result::Result<(), MigratableError>)>,
+    /// Event that the migration state should be triggered.
```
I find this sentence hard to understand. Please note that event is not a verb, but it reads like that here (at least to me).

Perhaps "Used to notify lister(s) to check migration results" or something like that would be better?

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2255,6 +2343,38 @@ impl Vmm {
+        let (vm, res) = receiver.recv().unwrap();
+
+        // at this point, the thread must be finished.
+        self.migration_thread_handle.take().unwrap().join().unwrap();
```
I am wondering why the receiver is needed here? Why can't the join handle just return the `Vm`?

I guess there is a good reason :slightly_smiling_face:

phip1611 replied:
actually, I think you are quite right and have a very good suggestion here :) thanks

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -3113,42 +3248,40 @@ impl RequestHandler for Vmm {
+                .spawn(move || {
+                    worker.run();
+                })
+                // just panic, as we're screwed when we can't spawn thready anyway
```
```suggestion
// just panic, as we're screwed when we can't spawn threads anyway
```

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2255,6 +2343,38 @@ impl Vmm {
+    ///
+    /// This should be called when the migration thread indicated a state#
+    /// change. The function parses the callback and either shuts down the VMM
+    /// on success or keeps the VM and the VMM running on migration failure.
```
```suggestion
    /// This should be called when the migration thread indicated a state
    /// change. The function parses the callback and either shuts down the VMM
    /// on success or keeps the VM and the VMM running on migration failure.
```

phip1611 replied:
thanks, good catch

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -786,6 +786,40 @@ pub struct VmmThreadHandle {
+pub enum MaybeVmOwnership {
+    /// The VMM holds the ownership of the VM.
+    Vmm(Vm),
+    /// The VM is temporarily by the current ongoing migration.
```
Can't you give suggestions like so?
```suggestion
/// The VM is temporarily blocked by the current ongoing migration.
```

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -786,6 +786,40 @@ pub struct VmmThreadHandle {
+impl MaybeVmOwnership {
+    /// Takes the VM and replaces it with [`Self::Migration`].
+    ///
+    /// Panics of `self` is not [`Self::Vmm`].
```
```suggestion
/// ## Panics 
/// This method panics if `self` is not [`Self::Vmm`].
```

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2753,41 +2802,51 @@ impl RequestHandler for Vmm {
-                    .as_ref()
-                    .map(|vm| vm.device_tree().lock().unwrap().clone());
+        // SAFETY: We know that the value is there.
+        let vm_config = unsafe { self.vm_config.as_ref().unwrap_unchecked() };
```
Can `vm_info` ever be in the hot path? If that is not the case then I would prefer a `.expect()` here.

phip1611 replied:
no, no hot path

phip1611 replied:
I replaced it with 


        let vm_config = self.vm_config.ok_or(VmError::VmNotCreated)?;
        let vm_config = vm_config.lock().unwrap().clone();

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2753,41 +2802,51 @@ impl RequestHandler for Vmm {
+
+        let device_tree = match &self.vm {
+            MaybeVmOwnership::Vmm(vm) => Some(vm.device_tree().lock().unwrap().clone()),
+            // TODO we need to fix this
+            MaybeVmOwnership::Migration => None,
```
Has this still not been addressed?

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2839,64 +2901,73 @@ impl RequestHandler for Vmm {
-                config.memory.size = desired_ram;
+        match self.vm {
+            MaybeVmOwnership::Vmm(ref mut vm) => {
+                if let Err(e) = vm.resize(desired_vcpus, desired_ram, desired_balloon) {
+                    error!("Error when resizing VM: {:?}", e);
+                    Err(e)
+                } else {
+                    Ok(())
+                }
```
```suggestion
vm.resize(desired_vcpus, desired_ram, desired_balloon).inspect_err(|e| error!("error when resizing VM: {:?}", e))
```

phip1611 replied:
fixed upstream

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2839,64 +2901,73 @@ impl RequestHandler for Vmm {
-                Ok(())
+        match self.vm {
+            MaybeVmOwnership::Vmm(ref mut vm) => {
+                if let Err(e) = vm.resize_zone(id, desired_ram) {
+                    error!("Error when resizing VM: {:?}", e);
+                    Err(e)
+                } else {
+                    Ok(())
+                }
```
```suggestion
vm.resize_zone(id, desired_ram).inspect_err(|e| error!("error when resizing VM: {:?}", e))
```

phip1611 replied:
This is already fixed upstream :) https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7482/commits/aad2995f00fc35a7b70d8f49e1a2a14acaaa20fb

So we  can leave it here as is

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2913,19 +2984,23 @@ impl RequestHandler for Vmm {
-            Ok(None)
+        match self.vm {
+            MaybeVmOwnership::Vmm(ref mut vm) => {
+                let info = vm.add_device(device_cfg).map_err(|e| {
+                    error!("Error when adding new device to the VM: {:?}", e);
+                    e
+                })?;
```
nit: I prefer `inspect_err` when the error itself doesn't need to change.

phip1611 replied:
good catch! I'll leave it as is here and fix it upstream.

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2942,39 +3017,49 @@ impl RequestHandler for Vmm {
-            Ok(None)
+        match self.vm {
+            MaybeVmOwnership::Vmm(ref mut vm) => {
+                let info = vm.add_user_device(device_cfg).map_err(|e| {
+                    error!("Error when adding new user device to the VM: {:?}", e);
+                    e
+                })?;
```
nit: Consider `inspect_err` when the error doesn't need to be modified.

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2942,39 +3017,49 @@ impl RequestHandler for Vmm {
-                Ok(())
+        match self.vm {
+            MaybeVmOwnership::Vmm(ref mut vm) => {
+                if let Err(e) = vm.remove_device(id) {
+                    error!("Error when removing device from the VM: {:?}", e);
+                    Err(e)
+                } else {
+                    Ok(())
+                }
```
```suggestion
vm.remove_device(id).inspect_err(|e| error!("error when removing device from the VM: {:?}, e))
```

phip1611 replied:
already fixed upstream https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7482/commits/aad2995f00fc35a7b70d8f49e1a2a14acaaa20fb

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2931,10 +2947,18 @@ impl RequestHandler for Vmm {
         match self.vm {
-            MaybeVmOwnership::Vmm(ref mut vm) => vm.resize_disk(id, desired_size),
+            MaybeVmOwnership::Vmm(ref mut vm) => {
+                if let Err(e) = vm.resize_disk(id, desired_size) {
+                    error!("Error when resizing disk: {:?}", e);
+                    Err(e)
+                } else {
+                    Ok(())
+                }
+            }
```
nit: if `vm.resize_disk` returns `Result<(),E>` then you can just use `.inspect_err` instead of pattern matching here.

phip1611 replied:
already fixed upstream https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7482/commits/aad2995f00fc35a7b70d8f49e1a2a14acaaa20fb

olivereanderson reviewed (commented):
This looks quite good,

I mostly have some nits and there are a few comments that I think can be improved.

I am wondering whether a design using the scoped thread API could also work here, but that does not need to be explored now.

phip1611 committed:
vmm: migration: prepare EventFd for async migration events

This is a pre-requisite for the following commit which puts the
migration into a dedicated thread. It allows the VMM to react to
migration events (success/failure).

The commit series was inspired by @ljcore [0] but was changed quite
significantly.

[0] https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7038

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
vmm: migration: handle in dedicated thread (make async)

This puts the send-migration action into a dedicated thread. This
means:

1. The send-migration call will exit sooner (just trigger the
   migration)
2. Other API Call will not be possible as the VM's ownership is
   transferred from the VMM to the migration thread. E.g., hotplugging
   won't work (which is good).
3. If the migration causes the VMM process to crash, this currently
   can't be observed. A mechanism to query the migration status doesn't
   exist.

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 commented:
@olivereanderson  since your review, I've simplified `VmMigrationGuard` and `MigrationThread` into a single struct. Further, the channel used to transfer the VM ownership is gone and now I return it via the JoinHandle. Thanks for the suggestion! Makes things very much simpler. LAst but not least, I've split the first commit into two. 

Code diff should be visible here: https://github.com/cyberus-technology/cloud-hypervisor/compare/6eab06b350aa0b504874a652735a328e23beaa21..937f8804a84ffb96ca8fcb9c2e8d279dde49933d

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2248,10 +2364,53 @@ impl Vmm {
+    ///
+    /// This should be called when the migration thread indicated a state
+    /// change. The function parses the callback and either shuts down the VMM
+    /// on success or keeps the VM and the VMM running on migration failure.
```
> The function parses the callback ...

Which callback?

Perhaps we have leftover documentation from an earlier version?

phip1611 replied:
yes, thanks!!

phip1611 replied:
fixed

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2248,10 +2364,53 @@ impl Vmm {
+    /// change. The function parses the callback and either shuts down the VMM
+    /// on success or keeps the VM and the VMM running on migration failure.
+    fn check_migration_result(&mut self) {
+        // At this point, the thread must be finished.
+        // TODO print nice error message and return error to outer world.
+        let (vm, migration_res) = self
+            .migration_thread_handle
+            .take()
+            .expect("should have thread")
+            .join()
+            .expect("should have joined");
```
Note that even with proper error handling we will loose the `Vm` if the thread panics. I guess we are assuming that can never be the case in which case the final `.expect` is alright.

phip1611 replied:
you are right - if we fail here, we are screwed anyway.

scholzp commented on vmm/src/lib.rs:
```diff
@@ -2681,15 +2873,18 @@ impl RequestHandler for Vmm {
-
-        self.vm_config = None;
-
         event!("vm", "deleted");
```
Should this follow the match clause below and not vice versa? What happens in case of `MaybeVmOwnership::Migration` below? Will the event then wrongfully be emitted?

phip1611 replied:
thanks, good catch!

scholzp reviewed (changes requested):
LGTM! Only one thing I have a question about.

amphi commented on vmm/src/lib.rs:
```diff
@@ -2248,10 +2364,54 @@ impl Vmm {
+                    let (sender, _) = &*ONGOING_LIVEMIGRATION;
+                    // unblock API call; propagate migration result
+                    sender.send(Err(e)).unwrap();
+                    sleep(Duration::from_millis(1));
```
Why is this necessary here but not when sending an `Ok()`?

phip1611 replied:
very good question. I think it is debugging leftover. I remove it

amphi commented on vmm/src/lib.rs:
```diff
@@ -2681,14 +2873,17 @@ impl RequestHandler for Vmm {
             return Ok(());
         }
 
-        // If a VM is booted, we first try to shut it down.
-        if self.vm.is_some() {
-            self.vm_shutdown()?;
-        }
-
-        self.vm_config = None;
+        match &self.vm {
+            MaybeVmOwnership::Vmm(_vm) => {
+                event!("vm", "deleted");
 
-        event!("vm", "deleted");
+                // If a VM is booted, we first try to shut it down.
+                self.vm_shutdown()?;
+                self.vm_config = None;
+            }
+            MaybeVmOwnership::Migration => Err(VmError::VmMigrating)?,
+            MaybeVmOwnership::None => {}
+        }
```
In the old version, `self.vm_config` gets deleted regardless of whether `self.vm` is some or not. But now you only delete `self.vm_config` if the VMM has ownership of the VM. This seems wrong, I guess you should also delete `self.vm_config` if the VM is not created yet.

phip1611 replied:
I gave it another shot. diff here: https://github.com/cyberus-technology/cloud-hypervisor/compare/697649ddceb8d8ee0a671d572e38187f0d1674a4..f9db2a3246f48234b90c3c513731e7c2cc5fef00

amphi reviewed (approved):
Looks good to me now.
